internals/secrethub/read_test.go

```diff
@@ -1,9 +1,134 @@
+
+	"github.com/secrethub/secrethub-cli/internals/cli/filemode"
+	"github.com/secrethub/secrethub-cli/internals/cli/ui/fakeui"
+	"github.com/secrethub/secrethub-go/internals/api"
+	"github.com/secrethub/secrethub-go/internals/assert"
+	"github.com/secrethub/secrethub-go/internals/errio"
+	"github.com/secrethub/secrethub-go/pkg/secrethub"
+	"github.com/secrethub/secrethub-go/pkg/secrethub/fakeclient"
 )
 
 func TestReadCommand_Run(t *testing.T) {
-	// TODO SHDEV-1029 Test ReadCommand.
+	testErr := errio.Namespace("test").Code("test").Error("test error")
+
+	cases := map[string]struct {
+		cmd            ReadCommand
+		newClientErr   error
+		versionService fakeclient.SecretVersionService
+		out            string
+		err            error
+	}{
+		"success read": {
+			cmd: ReadCommand{
+				path: "test/repo/secret",
+			},
+			versionService: fakeclient.SecretVersionService{
+				GetWithDataFunc: func(path string) (*api.SecretVersion, error) {
+					return &api.SecretVersion{
+						Data: []byte("testSecret"),
+					}, nil
+				},
+			},
+			out: "testSecret\n",
+		},
+		//"success clipboard": {
+		//	cmd: ReadCommand{
+		//		path:                "test/repo/secret",
+		//		clipper:             clip.NewClipboard(),
+		//		useClipboard:        true,
+		//		clearClipboardAfter: 5 * time.Minute,
+		//	},
+		//	versionService: fakeclient.SecretVersionService{
+		//		GetWithDataFunc: func(path string) (*api.SecretVersion, error) {
+		//			return &api.SecretVersion{
+		//				Data: []byte("testSecretClipboard"),
+		//			}, nil
+		//		},
+		//	},
+		//	out: "Copied test/repo/secret to clipboard. It will be cleared after 5 minutes.\n",
+		//},
+		"success file": {
+			cmd: ReadCommand{
+				path:     "test/repo/secret",
+				outFile:  "secret.txt",
+				fileMode: filemode.New(os.ModePerm),
+			},
+			versionService: fakeclient.SecretVersionService{
+				GetWithDataFunc: func(path string) (*api.SecretVersion, error) {
+					return &api.SecretVersion{
+						Data: []byte("testSecretFile"),
+					}, nil
+				},
+			},
+			out: "testSecretFile\n",
+		},
+		"fail file": {
+			cmd: ReadCommand{
+				path:     "test/repo/secret",
+				outFile:  "/fail/read.txt",
+				fileMode: filemode.New(os.ModeAppend),
+			},
+			versionService: fakeclient.SecretVersionService{
+				GetWithDataFunc: func(path string) (*api.SecretVersion, error) {
+					return &api.SecretVersion{
+						Data: []byte("testSecretFile"),
+					}, nil
+				},
+			},
+			out: "",
+			err: ErrCannotWrite("/fail/read.txt", "open /fail/read.txt: no such file or directory"),
+		},
+		"new client error": {
+			newClientErr: testErr,
+			err:          testErr,
+		},
+		"read error": {
+			versionService: fakeclient.SecretVersionService{
+				GetWithDataFunc: func(path string) (*api.SecretVersion, error) {
+					return nil, testErr
+				},
+			},
+			err: testErr,
+		},
+	}
+
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			// Setup
+			io := fakeui.NewIO(t)
+			tc.cmd.io = io
```

Comment:
Why make the separate `io` variable? Not only is it confusing (conflicts with the `io` package), it doesn't make sense to use it outside of the `tc.cmd` scope. Consider making it like this
```suggestion
			tc.cmd.io = fakeui.NewIO(t)
```

Reply:
The variable is made seperately because it is a mock and we can easily read the output that it has. I tried the suggestion you made, but I was no longer able to get the output's content. This happened due to the fact that the command accepts an interface, which means that I cannot use the `fakeui.FakeIO`'s specific fileds, in this case enabling me to see the output's content.
To solve this, I will rename the variable's name to avoid the conflict with the `io` package